Проверка на сабтаск выдёт 0, а не 1, не получается добавить сабтаск, …

AlexMarchz opened the pull request:
…не понимаю в чём дело

AlexMarchz committed:
Проверка на сабтаск выдёт 0, а не 1, не получается добавить сабтаск, не понимаю в чём дело

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,208 @@
+    private static final String HEADER = "id,type,name,description, epic\n";
+
+    public FileBackedTaskManager(File file) {
+        super(new InMemoryHistoryManager());
```
Лучше используй `Managers.getDefaultHistory()` при создании менеджера

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,208 @@
+            }
+            Files.createFile(file.toPath());
+        } catch (IOException io) {
+            throw new ManagerSaveException("Не удалось найти файл");
```
Каждый раз создавать и удалять файл не нужно, мы перезаписываем в один и тот же файл

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,208 @@
+        return fileManager;
+    }
+
+    public void save() {
```
Метод используется только в методах файлового менеджера, поэтому его стоит закрыть модификатором

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,208 @@
+            default:
+                return new Task(id, name, status, description);
+        }
+    }
```
Методы преобразования из строки/в строку лучше вынести в утилитный класс, чтобы разгрузить файловый менеджер

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,208 @@
+            default:
+                return new Task(id, name, status, description);
+        }
+    } */
```
Лишний код удаляй

Ivanov-s-tmn commented on src/manager/InMemoryTaskManager.java:
```diff
@@ -12,26 +12,31 @@
+    @Override
+    public int generatorId() {
+        return nextId++;
+    }
```
Этот метод не стоит открывать для публичного доступа, он модифицирует счетчик айди

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagertTest.java:
```diff
@@ -0,0 +1,77 @@
+
+        assertEquals(1, fileBackedTaskManager.tasks.size());
+        assertEquals(1, fileBackedTaskManager.epics.size());
+        assertEquals(0, fileBackedTaskManager.subTasks.size()); // не возвращается сабтаск
```
Нужно проверить айди эпика передаваемый при создании сабтаски `subTask = new SubTask("имяСабТаска", "описаниеСабТаска", Status.IN_PROGRESS, 2);`. Айди генерируется постфиксным инкрементом `return nextId++;` и эпику присваивается айди 1, а не 2

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagertTest.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Test
+    void testLoadFromFile() {
+        FileBackedTaskManager fileBackedTaskManager = new FileBackedTaskManager(file);
```
Инициализацию файлового менеджера тоже можно вынести в метод `@BeforeEach`

AlexMarchz committed:
конструктор FileBackedTaskManager использует Managers.getDefaultHistory();
создан класс TaskUtils для разгрузки FileBackedTaskManager;
убрано создание и удаление файла, теперь save() перезаписывает актуальный файл.

AlexMarchz committed:
конструктор FileBackedTaskManager использует Managers.getDefaultHistory();
создан класс TaskUtils для разгрузки FileBackedTaskManager;
убрано создание и удаление файла, теперь save() перезаписывает актуальный файл.

AlexMarchz committed:
CODESTYLE

AlexMarchz committed:
CODESTYLE

AlexMarchz committed:
CODESTYLE

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -68,70 +58,14 @@ public void save() {
-            default:
-                return "";
-        }
+        return TaskUtils.taskFromString(value);
```
Так можно просто вызывать где надо `TaskUtils.taskFromString()`, зачем дополнительные методы?)

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -3,20 +3,18 @@
 
     public FileBackedTaskManager(File file) {
-        super(new InMemoryHistoryManager());
+        super(Managers.getDefaultHistory()); //а почему так лучше?
```
Если вдруг у нас будет несколько реализаций менеджеров истории, то таким образом, мы можем редактировать только менеджер по умолчанию, который возвращает Managers и не придется менять реализацию везде вручную
